clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java

```diff
@@ -108,10 +94,16 @@ public long readFrom(ScatteringByteChannel channel) throws IOException {
                 }
             }
         }
-        if (buffer == null && requestedBufferSize != -1) { //we know the size we want but havent been able to allocate it yet
-            buffer = memoryPool.tryAllocate(requestedBufferSize);
-            if (buffer == null)
-                log.trace("Broker low on memory - could not allocate buffer of size {} for source {}", requestedBufferSize, source);
+        if (buffer == null && requestedBufferSize != -1) { //we know the size we want but haven't been able to allocate it yet
+            if (privileged) {
+                // If the channel is privileged, we allocate directly on the heap
+                buffer = ByteBuffer.allocate(requestedBufferSize);
+            } else {
+                buffer = memoryPool.tryAllocate(requestedBufferSize);
```

Comment:
Related to a previous question, but what happens when the requested buffer size exceeds the capacity of the pool?